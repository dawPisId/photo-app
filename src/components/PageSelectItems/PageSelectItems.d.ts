interface PageItemProps {
  currentPage: number;
  pageCount: number;
  setCurrentPage: (value: number) => void;
}
