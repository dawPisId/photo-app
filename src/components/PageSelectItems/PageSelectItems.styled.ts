import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-size: 1rem;
  svg {
    &.MuiSvgIcon-root {
      font-size: 1rem;
      cursor: pointer;
    }
  }
`;

export const SelectItem = styled.div`
  margin: 0.5rem;
  padding: 0.5rem 0.9rem;
  font-family: "Roboto", sans-serif;
  cursor: pointer;
  border-radius: 500px;
  &.active {
    background-color: lightgrey;
  }
  @media (max-width: 470px) {
    margin: 0.25rem;
  }
  @media (max-width: 400px) {
    margin: 0;
  }
  @media (max-width: 350px) {
    padding: 0.3rem 0.7rem;
  }
`;
