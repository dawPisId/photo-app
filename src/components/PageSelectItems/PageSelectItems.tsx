import { Container, SelectItem } from "./PageSelectItems.styled";
import { next, prev, range } from "../PageSelect/helpers";

import ArrowBackIosIcon from "@material-ui/icons/ArrowBackIos";
import ArrowForwardIosIcon from "@material-ui/icons/ArrowForwardIos";
import React from "react";

interface InputProps {
  currentPage: number;
  pageCount: number;
  setCurrentPage: (value:number) => void;
}
function PageSelectItems({ currentPage, setCurrentPage, pageCount }:InputProps) {
  let flags = {
    leftDots: false,
    rightDots: false,
    beginning: 0,
    end: 0,
  };
  switch (true) {
    case pageCount < 8: //no dots are displayed, pageCount is less or equal to item limit in selectBar.
      flags = {
        ...flags,
        beginning: 2,
        end: pageCount,
      };
      break;
    case currentPage <= 4 && pageCount > 7: //only right dots are displayed
      flags = {
        ...flags,
        rightDots: true,
        beginning: 2,
        end: 5,
      };
      break;
    case currentPage >= 5 && currentPage <= pageCount - 4 && pageCount > 7: //left and right dots are displayed
      flags = {
        leftDots: true,
        rightDots: true,
        beginning: currentPage - 1,
        end: currentPage + 1,
      };
      break;
    case currentPage >= pageCount - 3 && pageCount > 7: //only left dots are displayed
      flags = {
        ...flags,
        leftDots: true,
        beginning: pageCount - 4,
        end: pageCount - 1,
      };
      break;
    default:
      return <Container />;
  }
  let createPageNumbers = range(flags.beginning, flags.end).map(
    (page, index) => (
      <SelectItem
        key={index}
        className={currentPage === page ? "active" : undefined}
        onClick={() => setCurrentPage(page)}
      >
        {page}
      </SelectItem>
    )
  );
  return (
    <Container>
      <ArrowBackIosIcon onClick={() => prev(currentPage, setCurrentPage)} />
      <SelectItem
        className={currentPage === 1 ? "active" : undefined}
        onClick={() => setCurrentPage(1)}
      >
        1
      </SelectItem>
      {flags.leftDots && <SelectItem>...</SelectItem>}
      {createPageNumbers}
      {flags.rightDots && <SelectItem>...</SelectItem>}
      {pageCount > 7 && (
        <SelectItem
          className={currentPage === pageCount ? "active" : undefined}
          onClick={() => setCurrentPage(pageCount)}
        >
          {pageCount}
        </SelectItem>
      )}
      <ArrowForwardIosIcon
        onClick={() => next(currentPage, setCurrentPage, pageCount)}
      />
    </Container>
  );
}

export default PageSelectItems;
