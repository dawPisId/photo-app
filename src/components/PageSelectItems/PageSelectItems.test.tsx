import { fireEvent, render, screen } from "@testing-library/react";

import PageSelectItems from "./PageSelectItems";
import React from "react";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <PageSelectItems
        currentPage={10}
        pageCount={20}
        setCurrentPage={() => null}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("highlights to new page number on click", async () => {
  let tempValue = 10;
  const tempFunc = (value: number) => {
    tempValue = value;
  };
  const { rerender } = render(
    <PageSelectItems
      currentPage={tempValue}
      pageCount={20}
      setCurrentPage={tempFunc}
    />
  );
  fireEvent.click(screen.getByText(11));
  rerender(
    <PageSelectItems
      currentPage={tempValue}
      pageCount={20}
      setCurrentPage={tempFunc}
    />
  );
  expect(screen.getByText(11)).toHaveClass("active");
});
