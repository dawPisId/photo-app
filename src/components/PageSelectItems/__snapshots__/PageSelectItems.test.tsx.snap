// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bdnxRM kCPdri"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
    />
  </svg>
  <div
    className="sc-gtsrHT eihoZt"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="sc-gtsrHT eihoZt"
  >
    ...
  </div>
  <div
    className="sc-gtsrHT eihoZt"
    onClick={[Function]}
  >
    9
  </div>
  <div
    className="sc-gtsrHT eihoZt active"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="sc-gtsrHT eihoZt"
    onClick={[Function]}
  >
    11
  </div>
  <div
    className="sc-gtsrHT eihoZt"
  >
    ...
  </div>
  <div
    className="sc-gtsrHT eihoZt"
    onClick={[Function]}
  >
    20
  </div>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
    />
  </svg>
</div>
`;
